### ISSUE TITLE
WSL2 systemd RemoteApp connection invalid / WSLg apps not working

### ISSUE THREAD

**Issue by xcorp** [microsoft/WSL#8888]
### Version

Microsoft Windows [Version 10.0.22000.978]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.62.1-microsoft-standard-WSL2

### Distro Version

Ubuntu 20.04

### Repro Steps

Installed pre-release https://github.com/microsoft/WSL/releases/tag/0.67.6
Added 
`[boot]
systemd=true`
to /etc/wsl.conf

### Expected Behavior

Working WSLg apps with no alert windows.

### Actual Behavior

When I start WSL I get an alert windows with the text "The connection information for this resource is invalid"
![image](https://user-images.githubusercontent.com/876605/192214188-b4d42fc6-b2c1-4b56-93ca-481838e34c87.png)
Clicking OK closes the windows for a second before it appears again.

Starting a second  fresh WSL instance of another distro gives the same error, even though systemd=true is not set

Trying to start an app with WSLg, for example firefox doesn't work.
On the first WSL instance I get the following
```
$ firefox
Unable to init server: Could not connect: Connection refused
Error: cannot open display: :0

$ DISPLAY=:0 firefox
Unable to init server: Could not connect: Connection refused
Error: cannot open display: :0
```

Trying the same on the second instance firefox runs, but a window for it is never presented.

```
$ firefox
Error: no DISPLAY environment variable specified

$ DISPLAY=:0 firefox
[GFX1-]: glxtest: libEGL missing
[2022-09-26T07:25:12Z ERROR glean_core::metrics::ping] Invalid reason code startup for ping newtab
Missing chrome or resource URL: resource://gre/modules/UpdateListener.sys.mjs
^CExiting due to channel error.
```

Setting systemd to false on the first instance and running `wsl --shutdown`  restores all WSLg functionality and no alert windows show when starting WSL

**Comment by roblatham00** [comment]
I'm seeing different symptoms, but also graphics related so I'll add here (happy to open a new issue if that makes more sense).

When I enable systemd by editing /etc/wsl.conf some graphical applications are unable to load.
- xload: "Error: cannot open display: :0"
- gvim: "Error: cannot open display: :0"

However, some applications like the 'eog' image viewer and the 'evolution ' mail reader start up just fine.

Once I turn off systemd, xload and gvim work just fine.

**Comment by cerebrate** [comment]
@roblatham00 Your specific problem is probably that the X server socket is missing. Can you `ls /tmp/.X11-unix/X0`?

If not, you have the issue where systemd, on starting, clears out _/tmp_ and gets this socket on the way. Put [this file](https://gist.github.com/cerebrate/16a40d3c416d4f61f0a952d5f4033292) in /lib/tmpfiles.d and restart WSL to prevent this.

**Comment by benhillis** [comment]
@cerebrate - is there a way to update that symlink to be a bind mount instead?

**Comment by cerebrate** [comment]
@benhillis Not quite as easily as leveraging systemd-tmpfiles, since it doesn't have an option to create those. You'd have to have it create the mount point and then create a new systemd unit to run after systemd-tmpfiles to do the mounting.

**Comment by cerebrate** [comment]
@benhillis ...and since I am apparently constitutionally unable to _not_ build prototypes of things:

https://gist.github.com/cerebrate/5cc6b2d6b9a42812ced020ea3e59b059

does exactly that. 😁

**Comment by c4artisan** [comment]
> @benhillis ...and since I am apparently constitutionally unable to _not_ build prototypes of things:
> 
> https://gist.github.com/cerebrate/5cc6b2d6b9a42812ced020ea3e59b059
> 
> does exactly that. 😁

i did that and it didn't do anything. also broke the gnome applications

**Comment by cerebrate** [comment]
@c4artisan Let's keep that issue all in one place at #8899 .

**Comment by NotTheDr01ds** [comment]
@benhillis I've been meaning to post this as a separate issue for a while, but if you change the new bind mount (#8411 - good change by-the-way) to be read-only, I *think* (from my testing) it will prevent this from happening.  It looks like the Systemd-enablement changes are designed to inject some new `tmpfiles.d` files that *don't* overwrite the socket, but that doesn't seem to be 100% reliable.

According to the [Arch Systemd-nspawn documentation](https://wiki.archlinux.org/title/systemd-nspawn), which seems to be closely related to the WSL implementation, at one point at least this was required:

> **Note:** Since systemd version 235, `/tmp/.X11-unix` contents [have to be bind-mounted as read-only](https://github.com/systemd/systemd/issues/7093), otherwise they will disappear from the filesystem. The read-only mount flag does not prevent using `connect()` syscall on the socket. If you binded also /run/user/1000 then you might want to explicitly bind `/run/user/1000/bus` as read-only to protect the dbus socket from being deleted.

And a link directly to the [Poettering comment](https://github.com/systemd/systemd/pull/6979#issuecomment-336684161) (does he really work for MS now or is that just a rumor?) on the topic.

While there's some question as to whether or not this is still the case for Systemd-nspawn, I tried it soon after the bind-mount change, and (although I don't use Systemd all that often), it seemed to prevent `systemd-tmpfiles` from overwriting the WSLg socket.

**Comment by cerebrate** [comment]
@benhillis 

Still got this issue and can confirm it's because _tmp.mount_ is mounting a fresh tmpfs over the top of _tmp_ . Unmounting it manually lets me see the _/tmp/.X11-unix/X0_ WSLg bind-mounted there, but then I've also got an on-disk _/tmp_, so it's easier to just bind-mount it again myself.

**Comment by cerebrate** [comment]
Although whatever's doing the mounting of _/tmp/.X11-unix_ initially seems to have changed, as now it's causing my poor .mount unit to have conniptions. I've had to use a .service unit to do the re-bind-mounting and move it later in the systemd startup process (before _multi-user.target_ instead of before _sysinit.target_).

**Comment by benhillis** [comment]
@cerebrate - I'm a bit confused what's happening, we're not waiting for systemd to finish booting before setting up the /tmp/.X11-unix directory. It seems to be working locally for me.

**Comment by benhillis** [comment]
Ah I might have an idea of what this is, might you be running elevated / non elevated wsl.exe processes together?

**Comment by cerebrate** [comment]
@benhillis 

Oh! This is all a bit speculative, if you'll forgive me, but it seems to me that this all makes sense if it's a race condition.

If I'm right and your setting up of _/tmp/.X11-unix_ happens in parallel to the systemd startup, then depending on how fast systemd startup runs, either systemd will mount _/tmp_ first, _then_ the setup happens, and you end up with a working system with the _/tmp/.X11-unix_ mount visible; or _/tmp/.X11-unix_ will be setup first, then systemd will automatically mount _/tmp_ on top of it, and you end up in my scenario in which _/tmp/.X11-unix_ is empty because the mount point is hidden.

(The chronological ordering of /proc/mounts supports this, because mine - [gist here, as it's long](https://gist.github.com/cerebrate/d11aa7a52c02b32c2a512970ea7abb85) - clearly shows the second ordering of the mounts hiding the mount point.)

That would explain how some people get working systems and some don't, and how in the non-working case, unmounting the tmpfs _/tmp_ reveals the _/tmp/.X11-unix_ bind mount set up "underneath" it, in the underlying mount point directory.

Plausible?

**Comment by xcorp** [comment]
Just want to say that I launched a new instance and moved over what I needed from the old one. No issues so far with the new one.
Feel free to close issue.

**Comment by cerebrate** [comment]
I'd like to request we keep this issue open until the underlying problem's resolved, just to avoid having to create a new issue to track it.

**Comment by cerebrate** [comment]
@benhillis Sorry to say, I get the same mount ordering problem under 0.70.8; looks like the fix for #9038 didn't fix it for me.

**Comment by ThexXTURBOXx** [comment]
I am using `python3` through WSL 2 in PyCharm. I also get the same error (although in German) whenever PyCharm is running. When clicking on "OK" or closing the window, a new one opens.
This is really annoying and I cannot seem to find any solution other than reinstalling my WSL. Is there any other solution yet?

**Comment by wojtekow** [comment]
Same here, though the message is in Polish 😂

<img width="418" alt="image" src="https://user-images.githubusercontent.com/93520365/207803446-ce96b481-619b-4306-a3e3-b2eb62d80ec7.png">

**Comment by 4wk-** [comment]
Hello.

Same here: infinite error windows "RemoteApp", after waking up PC from sleep (I had multiple Debian launched)
`wsl genie -s`

```
wsl -l -v
  NAME      STATE           VERSION
* Debian    Running         2
  Ubuntu    Stopped         2
```

**Comment by romanofski** [comment]
I noticed that I get heaps of these windows with an open Explorer. Close the explorer and you can close the dialogs as well. I don't have information on the internals on what's going on here, but it's almost reliably reproducible with an open explorer window.

**Comment by ThexXTURBOXx** [comment]
@romanofski Not entirely true. The bug is reproducible when the process is open that uses WSL. In my case, this is not the explorer, but rather PyCharm. As soon as I close PyCharm, I can close the last dialog and no other one opens. The explorer was not open at that point.

**Comment by 4wk-** [comment]
Hey again. For the record, launching `wsl` with `genie` [is now outdated](https://devblogs.microsoft.com/commandline/systemd-support-is-now-available-in-wsl/), and I think this is related to this popin error. 

**Fix:** I completely uninstalled `wsl`, reinstalled it, then enabled systemD support the right way (without `genie`). 
https://gist.github.com/4wk-/889b26043f519259ab60386ca13ba91b

RemoteApp error is gone. Hope it helps someone!
